core/src/main/java/org/apache/iceberg/catalog/BaseSessionCatalog.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.catalog;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+
+public abstract class BaseSessionCatalog implements SessionCatalog {
+  private final Cache<String, Catalog> catalogs = Caffeine.newBuilder()
+      .expireAfterAccess(10, TimeUnit.MINUTES)
+      .removalListener((id, catalog, cause) -> {
+        if (catalog instanceof Closeable) {
+          try {
+            ((Closeable) catalog).close();
+          } catch (IOException e) {
+            throw new UncheckedIOException(String.format("Failed to close catalog: %s", catalog), e);
+          }
+        }
+      })
+      .build();
+
+  private String name = null;
+  private Map<String, String> properties = null;
+
+  @Override
+  public void initialize(String catalogName, Map<String, String> props) {
+    this.name = catalogName;
+    this.properties = props;
+  }
```

Comment:
Should we be inserting the catalog into the `catalogs` cache? I don't see where that is happening.

Reply:
The cache is creating the catalog if it doesn't exist when `get` is called.